functions/src/algoliaSync.ts

```diff
@@ -0,0 +1,125 @@
+import * as functions from 'firebase-functions'
+import * as algoliasearch from 'algoliasearch'
+import * as constants from './constants'
+import { algoliaMock } from './test/mocks'
+
+// Algolia Setup:
+const isTestMode = process.env.NODE_ENV === 'test'
+const client = isTestMode ? algoliaMock() : algoliasearch(constants.ALGOLIA_APP_ID, constants.ALGOLIA_SEARCH_API_KEY)
+
+// Triggered on create, update, and delete:
+export function activityDataHandler (change: functions.Change<FirebaseFirestore.DocumentSnapshot>, context: functions.EventContext) {
+  // @ts-ignore
+  return exports.algoliaSync<AlgoliaActivity>(client.initIndex(constants.ACTIVITY_INDEX_NAME), change, getAlgoliaActivityData, hasActivityChanged)
+}
+
+// Triggered on create, update, and delete:
+export function userDataHandler (change: functions.Change<FirebaseFirestore.DocumentSnapshot>, context: functions.EventContext) {
+  // @ts-ignore
+  return exports.algoliaSync<AlgoliaUser>(client.initIndex(constants.USER_INDEX_NAME), change, getAlgoliaUserData, hasUserChanged)
+}
+
+// Creates, updates, or deletes objects in the given Algolia index based on the
+// data from the document change. Utilizes dataFunc and compareFunc respectively
+// to get database data from change and to compare entries of type T.
+export async function algoliaSync<T extends AlgoliaBase>(algoliaIndex: algoliasearch.Index,
+                              change: functions.Change<FirebaseFirestore.DocumentSnapshot>,
+                              dataFunc: (doc: FirebaseFirestore.DocumentSnapshot) => T | null,
+                              compareFunc: (a: T, b: T) => boolean): Promise<void> {
+  const newData = dataFunc(change.after)
+  const oldData = dataFunc(change.before);
+  
+  // Create object:
+  if (!oldData && newData) {
+    await algoliaIndex.addObject(newData)
+  }
+
+  // Delete object:
+  if (!newData && oldData) {
+    await algoliaIndex.deleteObject(oldData.objectID)
+  }
+
+  // Update object:
+  if (newData && oldData && compareFunc(oldData, newData)) {
+    await algoliaIndex.partialUpdateObject(newData)
+  }
+}
+
+export function getAlgoliaUserData(doc: FirebaseFirestore.DocumentSnapshot): AlgoliaUser | null {
+  const data = doc.data()
+  if (!doc.exists || !data) {
+    return null
+  }
+  if (!data.favorite_topics || !data.blocked_users || !data.blocked_by || !data.location) {
+    return null
+  }
+  // Concatenates and removes duplicates:
+  const block_filter: string[] = [...new Set([].concat(data.blocked_by, data.blocked_users))]
+  return {
+    block_filter,
```

Comment:
no, this is _just_ for algolia - so we can easily check against one thing. Every time the user is updated in firestore with two arrays, it will combine to one for algolia